nibabel/parrec.py

```diff
@@ -990,6 +1019,54 @@ def get_rec_shape(self):
         inplane_shape = tuple(self._get_unique_image_prop('recon resolution'))
         return inplane_shape + (len(self.image_defs),)
 
+    def _strict_sort_keys(self):
+        """ Determine the sort order based on several image definition fields.
+
+        Data sorting is done in two stages:
+            - run an initial sort using several keys of interest
+            - call `vol_is_full` to identify potentially missing volumes
+              and add the result to the list of sort keys
+        """
+        # Sort based on a larger number of keys.  This is more complicated
+        # but works for .PAR files that get missorted by the above method
+        slice_nos = self.image_defs['slice number']
+        dynamics = self.image_defs['dynamic scan number']
+        phases = self.image_defs['cardiac phase number']
+        echos = self.image_defs['echo number']
+        image_type = self.image_defs['image_type_mr']
+
+        # try adding keys only present in a subset of .PAR files
+        idefs = self.image_defs
+        asl_keys = (idefs['label type'], ) if 'label type' in \
+            idefs.dtype.names else ()
+
+        if not self.general_info['diffusion'] == 0:
+            bvals = self.get_def('diffusion b value number')
+            if bvals is None:
+                bvals = self.get_def('diffusion_b_factor')
+            bvecs = self.get_def('gradient orientation number')
+            if bvecs is None:
+                # manually enumerate the different directions
+                bvecs = _direction_numbers(self.get_def('diffusion'))
+            diffusion_keys = (bvecs, bvals)
+        else:
+            diffusion_keys = ()
+
+        # Define the desired sort order (last key is highest precedence)
+        keys = (slice_nos, echos, phases) + \
+            diffusion_keys + asl_keys + (dynamics, image_type)
+
+        initial_sort_order = np.lexsort(keys)
+        is_full = vol_is_full(slice_nos[initial_sort_order],
+                              self.general_info['max_slices'])
+
+        # have to "unsort" is_full to match the other sort keys
```

Comment:
Do you mean the `is_full` case specifically?  If so, I don't think I tested that aside from verifying that it still passes the existing tests.  None of the new data sets I added have missing data, although I suppose we could truncate one of the files if desired.

I don't have an exhaustive set of tests for all available sort keys, but I did collect a few .PAR files with 4-6 dimensions to use as tests.  These are the new .PAR files I added in tests/data:

- *T1_dual_echo.PAR*:  A 4D data set:  (x, y, z, echo)
- *T1_3cho_mag_real_imag_phase.PAR*:  A 5D data set:  (x, y, z, echo, image_type)
- *ASL_3D_Multiecho.PAR*:  A 6D data set:  (x, y, z, echo, 'label type', dynamics)
The tests make sure the orders are sorted as expected for these cases.  I did not collect any data with cardiac phases, for example.

One other test I can think of would be to manually scramble the order of the entries in one of the .PAR files and make sure the sort keys that come out at the end still match the unscrambled case.